src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -3170,76 +3170,88 @@ exprt c_typecheck_baset::do_special_functions(
   }
   else if(
     identifier == "__builtin_add_overflow" ||
+    identifier == "__builtin_add_overflow_p")
+  {
+    return do_special_functions_builtin_overflow(expr, ID_plus);
+  }
+  else if(
     identifier == "__builtin_sub_overflow" ||
+    identifier == "__builtin_sub_overflow_p")
+  {
+    return do_special_functions_builtin_overflow(expr, ID_minus);
+  }
+  else if(
     identifier == "__builtin_mul_overflow" ||
-    identifier == "__builtin_add_overflow_p" ||
-    identifier == "__builtin_sub_overflow_p" ||
     identifier == "__builtin_mul_overflow_p")
   {
-    // check function signature
-    if(expr.arguments().size() != 3)
-    {
-      std::ostringstream error_message;
-      error_message << expr.source_location().as_string() << ": " << identifier
-                    << " takes exactly 3 arguments, but "
-                    << expr.arguments().size() << " were provided";
-      throw invalid_source_file_exceptiont{error_message.str()};
-    }
+    return do_special_functions_builtin_overflow(expr, ID_mult);
+  }
+  else
+    return nil_exprt();
+  // NOLINTNEXTLINE(readability/fn_size)
+}
 
-    typecheck_function_call_arguments(expr);
+exprt c_typecheck_baset::do_special_functions_builtin_overflow(
```

Comment:
Nit-pick: I'm a massive fan of moving this code out to a function :-) marginally less a fan of names such as "do_X" ;-) I'd personally lean towards `typecheck_special_functions_builtin_overflow` - its more verbose, yes, but just a bit more self-documenting, IMHO. Arguably the 'special_functions' part of the name could be dropped, leaving just `typecheck_builtin_overflow` ?

Reply:
I'm terrible at picking names, so I'm very happy to take up your advice here!